integration/test/TestLib/App.hs

```diff
@@ -0,0 +1,930 @@
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+
+-- |
+--
+-- Table of Contents:
+-- - SECTION_APP : App, Env, ServiceMap, Context
+-- - SECTION_ASSERTIONS
+-- - SECTION_JSON
+-- - SECTION_REQUEST : requests and responses
+-- - SECTION_PRINTING : Terminal and printingmodule App where
+-- - SECTION_MODSERVICE : Start services with modified configuration
+module TestLib.App where
+
+import Control.Concurrent.Async (mapConcurrently_)
+import Control.Exception (finally)
+import qualified Control.Exception as E
+import Control.Monad.Catch (MonadMask, MonadThrow)
+import Control.Monad.Catch.Pure (MonadCatch)
+import Control.Monad.Reader
+import Control.Monad.Trans.Maybe (MaybeT (..))
+import Control.Retry (fibonacciBackoff, limitRetriesByCumulativeDelay, retrying)
+import Data.Aeson hiding ((.=))
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as Aeson
+import qualified Data.Aeson.Key as KM
+import qualified Data.Aeson.KeyMap as KM
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as C8
+import Data.ByteString.Conversion
+import qualified Data.ByteString.Lazy as L
+import qualified Data.ByteString.Lazy.Char8 as LC8
+import qualified Data.CaseInsensitive as CI
+import qualified Data.Char as C
+import Data.List.Split (splitOn)
+import qualified Data.Map.Strict as Map
+import Data.Proxy (Proxy (Proxy))
+import qualified Data.Scientific as Sci
+import Data.Tagged
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Yaml as Yaml
+import GHC.Exception
+import GHC.Records
+import GHC.Stack
+import Imports hiding (ask, asks, local)
+import qualified Network.HTTP.Client as HTTP
+import qualified Network.HTTP.Types as HTTP
+import qualified Network.Socket as N
+import Network.URI (uriToString)
+import System.Exit (exitFailure)
+import System.FilePath (joinPath, takeDirectory, (</>))
+import System.FilePath.Posix (splitPath)
+import System.IO (hPutStrLn, openBinaryTempFile)
+import qualified System.IO.Error as Error
+import System.Process (CreateProcess (..), createProcess, proc, terminateProcess, waitForProcess)
+import Test.Tasty.Options
+import Test.Tasty.Providers
+import qualified Test.Tasty.Providers as Tasty
+import Test.Tasty.Providers.ConsoleFormat
+
+-------------------------------------------------------------------------------
+-- - SECTION_APP : App, Env, ServiceMap, Context
+-------------------------------------------------------------------------------
+
+newtype App a = App {unApp :: ReaderT Env IO a}
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadIO,
+      MonadMask,
+      MonadCatch,
+      MonadThrow,
+      MonadReader Env
+    )
+
+data AppFailure = AppFailure String
+
+instance Show AppFailure where
+  show (AppFailure msg) = msg
+
+instance Exception AppFailure where
+  displayException (AppFailure msg) = msg
+
+failApp :: String -> App a
```

Comment:
`Text` instead of `String` has my vote.  not for performance (I think that wouldn't be significant), but because it seems more natural to me.